Add controller rumble (#7)
v1993/cemuhook-protocol: README.md

| Offset | Length | Type | Meaning |
| ------ | ------ | ---- | ------- |
| 0 | 1 | Unsigned 8-bit | Is touch active (`1` if active, else `0`) |
| 1 | 1 | Unsigned 8-bit | Touch id (should be the same for one continuous touch) |
| 2 | 2 | Unsigned 16-bit | Touch X position |
| 4 | 2 | Unsigned 16-bit | Touch Y position |

## (Unofficial) Information about controller motors

> Ask/Reply how many motors in controller for some port.

Reviewer: ```suggestion
Query how many rumble motors specified controllers have.
```

### Incoming packet structure

| Offset | Length | Type | Meaning |
| ------ | ------ | ---- | ------- |
| 0 | 8 | Complex | Controller identification header, same as the incoming packet of "Actual controllers data" |

### Outgoing packet structure

| Offset | Length | Type | Meaning |
| ------ | ------ | ---- | ------- |
| 0  | 11  | Complex | Beginning described above |

> | 11 | 1 | Unsigned 8-bit | Motor count (0 means not support rumble, normally should be 2, and the motor id takes value from 0 to `count-1`) |

Reviewer: ```suggestion
| 11 | 1 | Unsigned 8-bit | Motor count - common values are 0 (no rumble support), 1 (single motor), and 2 (left/right motors) |
```

Single motor also sounds like a fairly common case.

## (Unofficial) Rumble controller motor

> Only incoming packet.

Reviewer: ```suggestion
Set rumble intensity for controller motor. Only incoming packet.

Clients should periodically (twice to ten times per second) re-send rumble state (both zero and non-zero) to account for potential packet loss. Servers should reset motor's rumble strength to zero if they haven't received rumble packets for it for prolonged period of time (whatever you use for client timeout is a good estimate, which should be about 5 seconds) to avoid controllers getting stuck rumbling in the event client disconnects disgracefully.
```

### Incoming packet structure

| Offset | Length | Type | Meaning |
| ------ | ------ | ---- | ------- |
| 0 | 8 | Complex | Controller identification header, same as the incoming packet of "Actual controllers data" |

> | 8 | 1 | Unsigned 8-bit | Motor id. |

Reviewer: ```suggestion
| 8 | 1 | Unsigned 8-bit | Motor id, 0~`motor count-1` |
```

| Offset | Length | Type | Meaning |
| ------ | ------ | ---- | ------- |
| 0 | 1 | Unsigned 8-bit | Is touch active (`1` if active, else `0`) |
| 1 | 1 | Unsigned 8-bit | Touch id (should be the same for one continuous touch) |
| 2 | 2 | Unsigned 16-bit | Touch X position |
| 4 | 2 | Unsigned 16-bit | Touch Y position |

## (Unofficial) Information about controller motors

Ask/Reply how many motors in controller for some port.

### Incoming packet structure

| Offset | Length | Type | Meaning |
| ------ | ------ | ---- | ------- |
| 0 | 8 | Complex | Controller identification header, same as the incoming packet of "Actual controllers data" |

For every controller motor you should send one packet structured like described below.

### Outgoing packet structure

| Offset | Length | Type | Meaning |
| ------ | ------ | ---- | ------- |
| 0  | 11  | Complex | Beginning described above |

> | 11 | 1 | Unsigned 8-bit | Motor id |

Reviewer: After thinking about it some more - how about just sending motor count per queried controller, with numbering going from zero? This has a pro of better handling situation with packet loss and allows to convey that client supports extension but lacks motors.

Author: OK, I have changed. The previous designed with reference to "Information about connected controllers" 😂

| Offset | Length | Type | Meaning |
| ------ | ------ | ---- | ------- |
| 0 | 1 | Unsigned 8-bit | Is touch active (`1` if active, else `0`) |
| 1 | 1 | Unsigned 8-bit | Touch id (should be the same for one continuous touch) |
| 2 | 2 | Unsigned 16-bit | Touch X position |
| 4 | 2 | Unsigned 16-bit | Touch Y position |

## (Optional) Information about controller motors

Ask/Reply how many motors in controller for some port.

> However, "DualShock" means there are two motors, so this interface does not need to be implemented. It can be agreed that id `0` is the left motor and id `1` is the right motor.

Reviewer: I don't think that making this optional to implement is a good idea for a few reasons:

1. While originally protocol was indeed designed only with DS3/4 in mind, it's used for many different controllers these days
2. Querying this interface should allow client to test if server supports rumble feature. As such, no response from this interface = no sending rumble commands to server that does not support them.

### Incoming packet structure

| Offset | Length | Type | Meaning |
| ------ | ------ | ---- | ------- |
| 0 | 1 | Unsigned 8-bit | Bitmask of actions you should take. Valid flags are `1` for slot-based registration, `2` for MAC-based registration, no bits (all set to `0`) to subscribe to all controllers. |
| 1 | 1 | Unsigned 8-bit | If slot-based registration is requested, slot to report about. |

> | 2 | 6 | **Unsigned 48-bit** | If MAC-based registration is requested, MAC of device to report about. |

Reviewer: With this table being copied a few times with next to no modifications it would make sense to describe it once in more generic terms instead ("Controller identification header").

For every controller motor you should send one packet structured like described below.

### Outgoing packet structure

| Offset | Length | Type | Meaning |
| ------ | ------ | ---- | ------- |
| 0  | 11  | Complex | Beginning described above |
| 11 | 1 | Unsigned 8-bit | Motor id |
| 12 | 1 | Unsigned 8-bit | Motor usable (`1` if usable, else `0`) |

> | 13 | 1 | N/A | Zero byte (`\0`). |

Reviewer: Any particular reason to add zero byte here? IIRC it's used to pad to a nice `2^n - 1` packet size in original packet.

Also, what's the point of reporting unusable motors?

## Rumble controller motor

Only incoming packet.

### Incoming packet structure

| Offset | Length | Type | Meaning |
| ------ | ------ | ---- | ------- |
| 0 | 1 | Unsigned 8-bit | Bitmask of actions you should take. Valid flags are `1` for slot-based registration, `2` for MAC-based registration, no bits (all set to `0`) to subscribe to all controllers. |
| 1 | 1 | Unsigned 8-bit | If slot-based registration is requested, slot to report about. |
| 2 | 6 | **Unsigned 48-bit** | If MAC-based registration is requested, MAC of device to report about. |
| 8 | 1 | Unsigned 8-bit | Motor id. |

> | 9 | 1 | Unsigned 8-bit | Motor state (`0` means no vibration, else means vibration intensity)|

Reviewer: IMO this note that intensity ranges from 0 to 255, just to be unambiguous.
